Allow item metadata to be expressed as attributes on the item element instead of child Xml elements

dsplaisted opened the pull request:
Resolves #708

AndyGerlicher commented:
Looks good so far, I like the approach! I'd like to see some more tests before we commit. Especially around adding elements that can't really be attributes (need to be escaped, etc.).

dsplaisted commented:
I've added some more tests and fixed some minor things.  I still need to add more tests around error cases, and we may need to add APIs to the project model in addition to the construction APIs.

dsplaisted commented:
I've added more tests and fixed some issues.  I think this is ready for review now.

jeffkl commented:
:shipit:

dsplaisted commented:
Anyone else want to review this one, or shall I go ahead and merge it?

rainersigwald commented:
I'll look this morning

jviau commented:
I think this PR is fine without the error rainersigwald brought up. MSBuild already supports 2 children of the same tag, taking the last one as the evaluated result. I think in this scenario the attribute can be considered before the child, so the a child metadata will win over an attribute.

For example, MSBuild supports this in the project file with `anything` being the result of `m1`:
```xml
<i1 Include="I">
  <m1>v1</m1>
  <m1>anything</m1>
</i1>
```

On a side note, this is a bit interesting. With the duplicate metadata you can retrieve the `ProjectItem` for that element, remove the `v1` metadata, and the project item will no longer have that metadata, but the `ProjectItemElement` does! So when you re-evaluted, the 1st metadata comes back to the project item. This is inconsistent, and either the remove should remove both metadata, or the project item should report the 1st metadata after the 2nd is removed without a re-evaluation. But this is a separate issue I feel.

dsplaisted commented on src/XMakeBuildEngine/Construction/ProjectElementContainer.cs:
```diff
@@ -455,6 +420,154 @@ protected internal virtual ProjectElementContainer DeepClone(ProjectRootElement
 
+        private void SetElementAsAttributeValue(ProjectElement child)
+        {
+            //  Assumes that child.ExpressedAsAttribute is true
```
Done

dsplaisted commented on src/XMakeBuildEngine/Construction/ProjectElementContainer.cs:
```diff
@@ -455,6 +420,154 @@ protected internal virtual ProjectElementContainer DeepClone(ProjectRootElement
+            string value = Microsoft.Build.Internal.Utilities.GetXmlNodeInnerContents(child.XmlElement);
+            ProjectXmlUtilities.SetOrRemoveAttribute(XmlElement, child.XmlElement.Name, value);
+
+            if (XmlElement.HasChildNodes)
```
Moved to the RemoveFromXml method and added comments

dsplaisted commented on src/XMakeBuildEngine/Construction/ProjectElementContainer.cs:
```diff
@@ -455,6 +420,154 @@ protected internal virtual ProjectElementContainer DeepClone(ProjectRootElement
+            }
+        }
+
+        //  If child "element" is actually represented as an attribute, update the value there
```
Done

dsplaisted commented on src/XMakeBuildEngine/Construction/ProjectElementContainer.cs:
```diff
@@ -455,6 +420,154 @@ protected internal virtual ProjectElementContainer DeepClone(ProjectRootElement
+            }
+            else
+            {
+                if (child.PreviousSibling != null)
```
I've added comments explaining this

dsplaisted commented on src/XMakeBuildEngine/Construction/ProjectItemElement.cs:
```diff
@@ -345,11 +345,17 @@ internal ProjectItemElement NextItem
+            return AddMetadata(name, unevaluatedValue, false);
+        }
+
+        public ProjectMetadataElement AddMetadata(string name, string unevaluatedValue, bool expressAsAttribute)
```
Done

dsplaisted commented on src/XMakeBuildEngine/Construction/ProjectMetadataElement.cs:
```diff
@@ -48,6 +48,14 @@ public string Name
         }
 
+        //  Add a new property with the same name here because this attribute should be public for ProjectMetadataElement,
+        //  but internal for ProjectElement, because we don't want it to be settable for arbintrary elements.
```
Done

dsplaisted replied:
Fixed

dsplaisted commented on src/XMakeBuildEngine/Evaluation/ProjectParser.cs:
```diff
@@ -368,6 +389,53 @@ private ProjectItemElement ParseProjectItemElement(XmlElementWithLocation elemen
+                }
+            }
+
+            //  Resorve attributes starting with underscores in case we need to add more built-in attributes later
```
Fixed

dsplaisted replied:
Pretty much, unless we're OK with potentially breaking some people.  Using an underscore was the best thing I could think of... if you have a better idea I'd love to hear it.

rainersigwald replied:
One concern: there's a common pattern of naming _properties_ with leading underscores as a clue to the user that they're internal implementation details. I don't think this is common for item metadata, though, so I don't think we'll be stepping on too many toes here.

The only other idea I have to make this even rarer is to take a page from the C++ book and use double underscores. Which is very ugly.

Thoughts, @AndyGerlicher @cdmihai @Sarabeth-Jaffe-Microsoft @jeffkl ?

dsplaisted commented on src/XMakeBuildEngine/Resources/Strings.resx:
```diff
@@ -1161,6 +1161,10 @@
     <comment>{StrBegin="MSB4173: "}</comment>
   </data>
+  <data name="InvalidAttributeDueToDuplication" UESanitized="false" Visibility="Public">
+    <value>MSB4227: The attribute &lt;{0}&gt; on element &lt;{1}&gt; is invalid because an attribute with that name already exists</value>
```
I'm matching the formatting here from the previous error message for a similar situation

rainersigwald replied:
All of the messages that I see referring to attributes use quotes (messages referring to elements use angle brackets). That makes sense to me.

dsplaisted replied:
Oh, I see what you mean.  I've updated it accordingly.

dsplaisted commented on src/XMakeBuildEngine/UnitTestsPublicOM/Construction/ProjectMetadataElement_Tests.cs:
```diff
@@ -77,6 +87,25 @@ public void ReadInvalidName()
+                            <i Include='i1' "+ "\u03A3" + @"='v1' />
+                        </ItemGroup>
+                    </Project>
+                ";
```
I've fixed this (though the pattern was just copied from the previous test).

dsplaisted commented:
As far as having a metadata attribute with the same name as a metadata element, I'd like to disallow it... but we'd have to add the check in several places (adding an attribute, changing its name, changing whether it's expressed as an attribute, parsing the XML).  It could potentially affect the perf.

So since this is a rare scenario where we already have some funky behavior, I'm leaning towards saying it's not worth adding the checks.

rainersigwald commented on src/XMakeBuildEngine/Construction/ProjectElementContainer.cs:
```diff
@@ -469,14 +467,19 @@ internal void AddToXml(ProjectElement child)
+                //  use the InsertAfter, InsertBefore, or AppendChild methods to add it in the right place.
+                //
+                //  Also, if PreserveWhitespace is true, then the elements we add won't automatically get indented, so
+                //  we try to match the surrounding formatting.
```
👍

rainersigwald commented on src/XMakeBuildEngine/Construction/ProjectElementContainer.cs:
```diff
@@ -423,30 +423,18 @@ protected internal virtual ProjectElementContainer DeepClone(ProjectRootElement
         }
 
-        //  If child "element" is actually represented as an attribute, update the value there
+        //  
```
ultra nit: extra blank comment line

dsplaisted replied:
Ultra fixed :-)

rainersigwald commented on src/XMakeBuildEngine/Construction/ProjectMetadataElement.cs:
```diff
@@ -49,7 +49,14 @@ public string Name
 
         //  Add a new property with the same name here because this attribute should be public for ProjectMetadataElement,
-        //  but internal for ProjectElement, because we don't want it to be settable for arbintrary elements.
+        //  but internal for ProjectElement, because we don't want it to be settable for arbitrary elements.
```
nit: is this a comment before the doc comments? should it just be in `<remarks>`?

dsplaisted replied:
It's talking about internal APIs, so I don't think it goes in the public API documentation that comes from the Xml comments

cdmihai commented on src/XMakeBuildEngine/Evaluation/ProjectParser.cs:
```diff
@@ -358,6 +356,27 @@ private ProjectItemElement ParseProjectItemElement(XmlElementWithLocation elemen
+                bool isKnownAttribute;
+                bool isValidMetadataNameInAttribute;
+
+                CheckMetadataAsAttributeName(attribute.Name, out isKnownAttribute, out isValidMetadataNameInAttribute);
```
Should this invalid attribute checking logic also be callable from the construction APIs? Otherwise API callers might set an invalid metadata name and only know about it later when the xml does not parse.

dsplaisted replied:
Good point, I've added checks to the construction APIs for this.

cdmihai commented:
Should the following construction APIs be update as well?
- [ProjectItem.SetMetadata](https://github.com/Microsoft/msbuild/blob/master/src/XMakeBuildEngine/Definition/ProjectItem.cs#L515)
- [ProjectRootElement.CreateMetadataElement](https://github.com/Microsoft/msbuild/blob/master/src/XMakeBuildEngine/Construction/ProjectRootElement.cs#L1525)

dsplaisted commented:
@cdmihai 

> Should the following construction APIs be updated as well?
> - [ProjectItem.SetMetadata](https://github.com/Microsoft/msbuild/blob/master/src/XMakeBuildEngine/Definition/ProjectItem.cs#L515)
> - [ProjectRootElement.CreateMetadataElement](https://github.com/Microsoft/msbuild/blob/master/src/XMakeBuildEngine/Construction/ProjectRootElement.cs#L1525)

I don't think they need to be.  `ProjectRootElement.CreateMetadataElement` returns a `ProjectMetadataElement`, which you can set the `ExpressedAsAttribute` property on.  Likewise, `ProjectItem.SetMetadata` returns a `ProjectMetadata`, which has an `Xml` property of type `ProjectMetadataElement`, which you can set the `ExpressedAsAttribute` property on.

jviau commented:
@rainersigwald yes this is fine for CPS.
